lib/TPP/ConvertLinalgToTpp.cpp

```diff
@@ -35,31 +35,55 @@ struct ConvertGenericOpToTpp : public OpRewritePattern<linalg::GenericOp> {
   LogicalResult rewriteToTppOp(linalg::GenericOp linalgOp,
                                PatternRewriter &rewriter) const {
     SmallVector<Value> operands;
+    if (tpp::utils::isTppZero(linalgOp, &operands)) {
+      assert((operands.size() != 0 && operands.size() <= 2) &&
+             "zero expects one or two operands");
+
+      // Only take the output, the other operand might be a constant.
+      rewriter.replaceOpWithNewOp<tpp::ZeroOp>(linalgOp,
+                                               operands[operands.size() - 1],
+                                               operands[operands.size() - 1]);
```

Comment:
I'm assuming one op is at tensor level and two at memref level, but both have to be the same.

Here, we ignore the first, without checking it. What if they're different? If they verify forbids, I still think we should add an `assert` and help developers see that it's an error.

Also, if they're identical, just taking the `[0]` element would do, without the need to have `size - 1`.

Reply:
At this stage we're looking at `linalg.generic` operands not `tpp.zero` operands.
This takes into account the `@zero_fill_arg` test case (`linalg-to-tpp-memref.mlir`) where the two `linalg.generic` operands are: lhs a zero constant `arith.constant 0.0` and rhs a `memref`.
In case like `@zero_fill_const` only one `linalg.generic` operand is returned that is a `memref` from its `outs`. So we need to only take the output which will always be the last item in `operands`.

Reply:
I get it but we still need to make sure they're the same, as the semantics isn't the same if not.

I also think the assert above should really not be there. 

It should just not be a match.

Reply:
But the operands of `linalg.generic` don't have to be the same for to represent zero fill. As long as we fill a 1D or a 2D buffer with zero values that is equivalent to a `tpp.zero`.

Reply:
We allow similar case with relu it's just this operand matching complexity is hidden somewhere else.
I'll just shift it to TppUtils as it is for relu to make this conversion logic cleaner.

Reply:
In the examples you have, the output memref is always the last one. What if it isn't? What if there are two inputs? Or two outputs? You still have two operands and you'll always take the last. The code you wrote matches your tests, yes, but it can also match other cases that are not `tpp.zero`.

At the very least, it passes the information that it's "the last operand" that is relevant, when actually, it's "the destination memref". If we make that clear, it's not only harder to match the wrong thing here, but it also helps developers to understand what the constraints are.

My suggestion is to:
1. move all checks to match a `tpp.zero` inside `isTppZero`, including all asserts and op checks.
2. create a `tpp.zero` with the operand that is **guaranteed** to be the output memref.

Reply:
> What if there are two inputs? Or two outputs?

Then it fails the definition of a tpp unary op. Or the definition of an tpp op all together if it is really weird generic.
But I agree here that it is unclear at the conversion level. These assumptions are tied to `isTppZero`, I'll shift all the validation logic there.